### PR TITLE
Prepopulate ResourceType Names in Tenant Allocation #811

### PR DESCRIPTION
**Pull request by prb112**
Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java [hunk 1]
```diff
@@ -816,6 +817,23 @@ protected void allocateTenant() {
         // Fill any static data tables (which are also partitioned by tenant)
         populateStaticTables(gen, tenantKey);
 
+        // Prepopulate the Resource Type Tables
+        try (ITransaction tx = TransactionFactory.openTransaction(connectionPool)) {
+            try {
+                Connection c = connectionPool.getConnection();
```
Comment (albertwang-ibm) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java]: can we use one try here?

**Comment by lmsurpre** [comment]
I've always wondered same thing!  Why are we building up this static list dynamically?!

I think it would be even better if we could ensure that the RESOURCE_NAMES table is the same for each tenant.  That is, each tenant should have the same ids for each resource type (instead of letting the db assign them using a SEQUENCE like it does now).

**Comment by prb112** [comment]
I'll also look at updating the Java code to optimize (instead of retrieving everytime)

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/Db2PopulateResourceTypes.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                try (CallableStatement cStmt = c.prepareCall(callableStr);) {
+                    cStmt.setString(1, resourceType);
+                    cStmt.registerOutParameter(2, Types.INTEGER);
+                    cStmt.execute();
+                }
+            }
+        } catch (SQLException x) {
+            throw translator.translate(x);
+        }
+
```
Comment (albertwang-ibm) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/Db2PopulateResourceTypes.java]: better to remove the empty lines.

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/Db2PopulateResourceTypes.java]: will do.

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java (hunk 1 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java]: it's the pattern that Robin follows through the code. thus the transaction factory, and then the nested tries

Comment (albertwang-ibm) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java]: not sure if we need to close the connection or put it back to the connection pool ...

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/Db2PopulateResourceTypes.java (hunk 2 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/Db2PopulateResourceTypes.java]: fixed in my local commit

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java (hunk 1 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java]: We do not need to put it back in the pool... from what I've reviewed in the code, however, you are correct that we need to do autocloseable.

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java]: fixed

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/Db2PopulateResourceTypes.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                // Check Error Codes. 
+                int[] codes = batch.executeBatch();
+                int errorCodes = 0;
+                for (int code : codes) {
+                    if (code != 0) {
+                        errorCodes++;
+                    }
+                }
+                if (errorCodes > 0) {
+                    LOGGER.warning("at least one of the Resource Types are not populated [" + errorCodes + "]");
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/Db2PopulateResourceTypes.java]: maybe it should be `Level.SEVERE` and rollback?  Is there a case where we'd see this and want to keep going?

Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/Db2PopulateResourceTypes.java]: You are right - fixing

**Comment by prb112** [comment]
Fixed

**Pull request merged**